### PR TITLE
{Packaging} Bump cryptography to 38.0.4, pyOpenSSL to 22.1.0

### LINKED ISSUE #24780: Bump cryptography and pyOpenSSL to make CLI compatible with openssl3
Current version is not compatible with openssl3. We should do this before other OS use openssl3 as default version.

Once it's done, https://github.com/Azure/azure-cli/pull/24768 could be reverted.

### PR DESCRIPTION
**Pull request by benjamin-brady**
Close #24780.
This also fixes "Failed building wheel for cryptography" error on Mac.

**Related command**
Installing packages though `azdev setup`

**Description**
This change fixes a `Could not build wheels for cryptography` error which occurs when installing packages on an M1 Mac via `azdev setup`.

**How to reproduce**
Install prerequisites:
```
pip3 install azdev
git clone https://github.com/Azure/azure-cli-extensions
cd ./azure-cli-extensions
```
Run azdev setup, which will fail when installing azure-cli darwin packages.
```
azdev setup --repo ./  --cli ../azure-cli

=======================
| Azure CLI Dev Setup |
=======================

Azure CLI:
    /Users/benbrady/Documents/azure-cli

Azure CLI extension repos:
    /Users/benbrady/Documents/azure-cli-extensions-r2k1

 Installing packages 
=====================

Upgrading pip...
Installing `requirements.txt`...
Installing `azure-cli-telemetry`...
Installing `azure-cli-core`...
Installing `azure-cli`...
Installing `azure-cli-testsdk`...
Installing `requirements.py3.Darwin.txt`...
Command `/Users/benbrady/Documents/azure-cli-extensions-r2k1/env/bin/python -m pip install -r /Users/benbrady/Documents/azure-cli/src/azure-cli/requirements.py3.Darwin.txt` failed with exit code 1:

...

      clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -I/Users/benbrady/Documents/azure-cli-extensions-r2k1/env/include -I/Users/benbrady/.pyenv/versions/3.10.8/include/python3.10 -c build/temp.macosx-13.0-arm64-cpython-310/_openssl.c -o build/temp.macosx-13.0-arm64-cpython-310/build/temp.macosx-13.0-arm64-cpython-310/_openssl.o -Wconversion -Wno-error=sign-conversion
      build/temp.macosx-13.0-arm64-cpython-310/_openssl.c:575:10: fatal error: 'openssl/opensslv.h' file not found
      #include <openssl/opensslv.h>
               ^~~~~~~~~~~~~~~~~~~~
      1 error generated.
      
          =============================DEBUG ASSISTANCE=============================
          If you are seeing a compilation error please try the following steps to
          successfully install cryptography:
          1) Upgrade to the latest pip and try again. This will fix errors for most
             users. See: https://pip.pypa.io/en/stable/installing/#upgrading-pip
          2) Read https://cryptography.io/en/latest/installation.html for specific
             instructions for your platform.
          3) Check our frequently asked questions for more information:
             https://cryptography.io/en/latest/faq.html
          =============================DEBUG ASSISTANCE=============================
      
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for cryptography
Failed to build cryptography
ERROR: Could not build wheels for cryptography, which is required to install pyproject.toml-based projects

```

**Testing Guide**
On an *M1 Mac* run the following.
```
pip3 install azdev
git clone https://github.com/Azure/azure-cli-extensions
cd ./azure-cli-extensions
azdev setup --repo ./  --cli ../azure-cli
```
Observe darwin packaged are installed successfully with no errors.
```
...
Installing `requirements.py3.Darwin.txt`...

Elapsed time: 3 min 2 sec

 Finished dev setup! 
=====================
```

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Misc

**Comment by bebound** [comment]
Thanks for your contribution.
We use the same package on different platform, could you please also change requirements for Linux and Window?

**Comment by benjamin-brady** [comment]
I've bumped the packages on all platforms. Please take a look.

**Comment by bebound** [comment]
@jiasli  @evelyn-ys for review

**Review by jiasli — APPROVED** [review]

**Review by bebound — APPROVED** [review]

**Comment by benjamin-brady** [comment]
Please review again as I've resolved some conflicts.

**Review by bebound — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

File: src/azure-cli/requirements.py3.windows.txt [hunk 1]
```diff
@@ -125,8 +125,8 @@ PyGithub==1.55
 PyJWT==2.4.0
 PyMySQL==1.0.2
 PyNaCl==1.5.0
+pyOpenSSL==22.1.0
 pyodbc==4.0.35
```
Comment (bebound) [inline src/azure-cli/requirements.py3.windows.txt:129]:
```suggestion
pyodbc==4.0.35
pyOpenSSL==22.1.0
```
It should be alphabetical order.

**Comment by benjamin-brady** [comment]
Please feel free to merge this as I don't have authorization.

**Review by bebound — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

**Comment by benjamin-brady** [comment]
Thanks @bebound and @evelyn-ys 🙏. Could one of you please merge this on my behalf as I don't have permissions?

**Comment by jiasli** [comment]
FYI: In `src/azure-cli-core/setup.py`, previously we set an upper bound for `cryptography` as `'cryptography>=3.2,<3.4'` because

> `3.4+` versions rely on rust and cause install issues on some platforms.

See: https://github.com/Azure/azure-cli/pull/15687#discussion_r578898876

Later the constraint were loosened by https://github.com/Azure/azure-cli/pull/19639, so that community packagers can choose whichever version of `cryptography` they want to use.

Now as we only bump `cryptography` in `src/azure-cli/requirements.py3.*.txt`, as long as CI passes and the built package works, we can merge this PR.

Users using `pip` to install Azure CLI or relying `setup.py` should be responsible for choosing the compatible `cryptography` package version and tackling security vulnerabilities in older versions of `cryptography`.

**Review by jiasli — APPROVED** [review]

**Comment by jiasli** [comment]
Ops, looks like we have 39.0.0 now:

https://pypi.org/project/cryptography/

image

**Pull request merged**